daemon/policy.go

```diff
@@ -276,7 +277,41 @@ func (d *Daemon) PolicyDelete(labels labels.LabelArray) *apierror.APIError {
 		return apierror.New(DeletePolicyNotFoundCode, "policy not found")
 	}
 
-	d.TriggerPolicyUpdates([]policy.NumericIdentity{})
+	go func() {
+		wg := d.TriggerPolicyUpdates([]policy.NumericIdentity{})
+		// Wait for all policies to be updated so that
+		// we can grab a fresh map of consumables.
+		wg.Wait()
+
+		consumables := d.consumableCache.GetConsumables()
+
+		endpointmanager.Mutex.RLock()
+		for _, ep := range endpointmanager.Endpoints {
+			ep.Mutex.RLock()
+			// If the policy is not being enforced then keep the CT
+			// entries
+			// FIXME: Do we really need to check the daemon.PolicyEnforcement()?
+			if ep.SecLabel == nil ||
+				!ep.Opts.IsEnabled(endpoint.OptionPolicy) ||
+				d.PolicyEnforcement() == endpoint.NeverEnforce {
+				ep.Mutex.RUnlock()
+				continue
+			}
+			epSecID := ep.SecLabel.ID
+
+			ep.Mutex.RUnlock()
+
+			idsToKeep := map[uint32]bool{}
+			if consumers, ok := consumables[epSecID]; ok {
+				for _, consumer := range consumers {
+					idsToKeep[consumer.Uint32()] = true
+				}
+			}
+			log.Debugf("Keeping entries of EP %d: %+v", ep.ID, idsToKeep)
+			endpointmanager.KeepCTEntryOf(!d.conf.IPv4Disabled, ep, idsToKeep)
```

Comment:
In your example, I'm not worried about new CT entries. Your case "2" has to depend on an existing allow consumer so it will be fine.

What I'm worried about is a new security identity popping up while we generate the `IDsToKeep` list. The new security identity is properly allowed in the policymap and a TCP connection is established. This will then get dropped.

Could we instead create the list of identities which we have to *remove*? This would make this race condition impossible.